sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala

```diff
@@ -155,3 +159,169 @@ case class GpuShuffledHashJoinExec(
     if (isSkewJoin) super.nodeName + "(skew=true)" else super.nodeName
   }
 }
+
+object GpuShuffledHashJoinExec extends Arm {
+  /**
+   * Helper iterator that wraps a BufferedIterator of AutoCloseable subclasses.
+   * This iterator also implements AutoCloseable, so it can be closed in case
+   * of exceptions.
+   *
+   * @param wrapped the buffered iterator
+   * @tparam T an AutoCloseable subclass
+   */
+  class CloseableBufferedIterator[T <: AutoCloseable](wrapped: BufferedIterator[T])
+    extends BufferedIterator[T] with AutoCloseable {
+    override def head: T = wrapped.head
+    override def headOption: Option[T] = wrapped.headOption
+    override def next: T = wrapped.next
+    override def hasNext: Boolean = wrapped.hasNext
+    override def close(): Unit = {
+      headOption.foreach(_.close())
+    }
+  }
+
+  /**
+   * Gets a `ColumnarBatch` and stream Iterator[ColumnarBatch] pair by acquiring
+   * the GPU semaphore optimally in the scenario where the build side is relatively
+   * small (less than batch goal).
+   *
+   * In the optimal case, this function will load the build side on the host up to the
+   * goal configuration and if it fits within the goal, allow the stream iterator
+   * to also pull to host its first batch. After the first batch is on the host, the
+   * stream iterator acquires the semaphore and then the build side is copied to the GPU.
+   *
+   * Prior to this we would get a build batch on the GPU, acquiring
+   * the semaphore in the process, and then begin pulling from the stream iterator,
+   * which could include IO (while holding onto the semaphore).
+   *
+   * The function handles the case where the build side goes above the configured batch
+   * goal, in which case it will concat on the host, grab the semaphore, and continue to
+   * pull the build iterator to build a bigger batch on the GPU. This is not optimized
+   * because we hold onto the semaphore during the entire time after realizing the goal
+   * has been hit.
+   *
+   * @param targetBatchSize configured target batch size
+   * @param buildOutput output attributes of the build plan
+   * @param buildIter build iterator
+   * @param streamIter stream iterator
+   * @param semWait wait GPU metric
+   * @param buildTime build GPU metric
+   * @return a pair of `ColumnarBatch` and streamed iterator that can be
+   *         used for the join
+   */
+  def getBuiltBatchAndStreamIter(targetBatchSize: Long,
```

Comment:
We cannot as any of the iterators if they `haveNext` unless we hold the GPU semaphore or we know 100% that they will not put something on the GPU.  I the general case we cannot know that. So we cannot call it on the stream iterator and we cannot call it on the build iterator to discover what type of iterator it is.

Reply:
I might not understand how/when the semaphore is grabbed in these cases. If I am wrong, which I hope I am, then we need some more comments explaining how this all works.

Reply:
Ya I rethought it. Can we have some comments around the hasNext calls saying that they will grab the semaphore if they are going to put something on the GPU and we don't need to worry about it in those cases.

Reply:
Yeah, if hasNext is called in the buffered iterator a single batch is getting loaded. That batch may be on the GPU. The expectation is that the iterator that gave us that batch via the buffered iterator is getting on the GPU, if not that's a bug. I'll put a comment around this to make this clear.